src/Tools/AnalyzerRunner/Program.cs

```diff
@@ -0,0 +1,480 @@
+            Parallel.ForEach(fixes, fix =>
+            {
+                try
+                {
+                    fix.GetOperationsAsync(cancellationToken).GetAwaiter().GetResult();
+                }
+                catch (Exception ex)
+                {
+                    // Report thrown exceptions
+                    lock (lockObject)
+                    {
+                        Utilities.WriteLine($"The fix '{fix.Title}' threw an exception:", ConsoleColor.Yellow);
+                        Utilities.WriteLine(ex.ToString(), ConsoleColor.Red);
+                    }
+                }
+            });
+
+            Console.WriteLine($"Calculating changes completed in {stopwatch.ElapsedMilliseconds}ms");
+        }
+
+        private static async Task<IEnumerable<CodeAction>> GetFixesAsync(Solution solution, CodeFixProvider codeFixProvider, Diagnostic diagnostic, CancellationToken cancellationToken)
+        {
+            List<CodeAction> codeActions = new List<CodeAction>();
+
+            await codeFixProvider.RegisterCodeFixesAsync(new CodeFixContext(solution.GetDocument(diagnostic.Location.SourceTree), diagnostic, (a, d) => codeActions.Add(a), cancellationToken)).ConfigureAwait(false);
+
+            return codeActions;
+        }
+
+        private static Task<Statistic> GetAnalyzerStatisticsAsync(IEnumerable<Project> projects, CancellationToken cancellationToken)
+        {
+            ConcurrentBag<Statistic> sums = new ConcurrentBag<Statistic>();
+
+            Parallel.ForEach(projects.SelectMany(i => i.Documents), document =>
+            {
+                var documentStatistics = GetAnalyzerStatisticsAsync(document, cancellationToken).ConfigureAwait(false).GetAwaiter().GetResult();
+                sums.Add(documentStatistics);
+            });
+
+            Statistic sum = sums.Aggregate(new Statistic(0, 0, 0), (currentResult, value) => currentResult + value);
+            return Task.FromResult(sum);
+        }
+
+        private static async Task<Statistic> GetAnalyzerStatisticsAsync(Document document, CancellationToken cancellationToken)
+        {
+            SyntaxTree tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+
+            SyntaxNode root = await tree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var tokensAndNodes = root.DescendantNodesAndTokensAndSelf(descendIntoTrivia: true);
+
+            int numberOfNodes = tokensAndNodes.Count(x => x.IsNode);
+            int numberOfTokens = tokensAndNodes.Count(x => x.IsToken);
+            int numberOfTrivia = root.DescendantTrivia(descendIntoTrivia: true).Count();
+
+            return new Statistic(numberOfNodes, numberOfTokens, numberOfTrivia);
+        }
+
+        private static IEnumerable<DiagnosticAnalyzer> FilterAnalyzers(IEnumerable<DiagnosticAnalyzer> analyzers, Options options)
+        {
+            foreach (var analyzer in analyzers)
+            {
+                if (options.UseAll)
```

Comment:
there seems no way to filter analyzers to just 1 analyzer except DiagnosticIds. are we asking users to know diangositcs ID the analyzer they are looking for is supporting? can we just add support for users explicitly specific analyzer name they want to run?

Reply:
:memo: The current approach is a carry-over from StyleCopTester. I see two options here:

1. Merge with this form, and update in the next PR to support specifying analyzers by name
2. Fix in this PR

Since the feature is very important for our analyzers (which often have many IDs and have changed over time), I would generally lean towards 2. However, I think we could go with 1 and the review process would be simplified. Option 1 also allows us to discuss the specific command line option we want to add for this.

Reply:
replaced with names

---
In reply to: [149861556](https://github.com/dotnet/roslyn/pull/23087#discussion_r149861556) [](ancestors = 149861556)